### PR TITLE
Add state filter to dag_runs endpoint

### LINKED ISSUE #16844: Rest API: allow filtering DagRuns by state.
**Add state filter to the dag runs list API endpoint**

One feature available in the "Browse / Dag Runs" page but not in the current rest API is the ability to filter runs of a specific state(s).

Example use-case: this would let a client efficiently fetch the number of "queued" and "running" runs, or look at recent failed runs.

Ideally, the current `/dags/{dag_id}/dagRuns` and `/dags/~/dagRuns/list` endpoints would each get updated to support an additional parameter called `state`. This parameter could be given multiple times and act as a logical "OR" (just like `dag_ids` in the POST endpoint, or like `state` in the task instances endpoint).

The web UI page offers more fancy filters like "includes", but for something with a finite number of values like `state`, it doesn't feel necessary for the API.

### PR DESCRIPTION
**Pull request by gyscos**
This adds a new `state` filter parameter to the dagRuns endpoints.

closes: #16844

### CODE REVIEW

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 1]
```diff
@@ -125,15 +129,16 @@ def get_dag_runs(
...
+    end_date_gte: Optional[str],
+    end_date_lte: Optional[str],
+    execution_date_gte: Optional[str],
+    execution_date_lte: Optional[str],
+    start_date_gte: Optional[str],
+    start_date_lte: Optional[str],
+    state: Optional[List[str]],
+    limit: Optional[int],
+    offset: Optional[int],
+    order_by: str,
```
Comment (gyscos) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]: I added type annotations to the functions I affected, but did not touch the others.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 2]
```diff
@@ -144,11 +149,14 @@ def _fetch_dag_runs(
...
     )
+
+    query = apply_array_filter(query, key=DagRun.state, values=state)
+
     # Count items
     total_entries = query.count()
     # sort
     to_replace = {"dag_run_id": "run_id"}
-    allowed_filter_attrs = [
+    allowed_sort_attrs = [
```
Comment (gyscos) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py:167]: Renamed as the previous name was slightly misleading, as it was not filtering anything.

Comment (uranusjr) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:167]: Likely a copy-paste error, thanks for cleaning this up.

File: airflow/api_connexion/parameters.py [hunk 3]
```diff
@@ -89,6 +89,13 @@ def wrapped_function(*args, **kwargs):
     return format_parameters_decorator
 
 
+def apply_array_filter(query, key, values):
```
Comment (uranusjr) [inline airflow/api_connexion/parameters.py]: Could you also add type annotation to this function?

File: airflow/api_connexion/endpoints/task_instance_endpoint.py [hunk 4]
```diff
@@ -120,7 +114,7 @@ def get_task_instances(
     end_date_lte: Optional[str] = None,
     duration_gte: Optional[float] = None,
     duration_lte: Optional[float] = None,
-    state: Optional[str] = None,
+    state: Optional[List[str]] = None,
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/task_instance_endpoint.py]: So the previous annotation was wrong, right? The value passed would would be `List[str]` even before this change, this simply fixes the type annotation to match reality, from what I can tell.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 5]
```diff
@@ -218,6 +226,7 @@ def get_dag_runs_batch(session):
         data["execution_date_lte"],
         data["start_date_gte"],
         data["start_date_lte"],
+        date["state"],
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
        data["state"],
```

Oops

**Review by uranusjr — CHANGES_REQUESTED** [review]
You got a typo in the implementation, otherwise this is awesome, many thanks!

File: airflow/api_connexion/endpoints/dag_run_endpoint.py (hunk 5 above)
Comment (gyscos) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py]: Whoops indeed

File: airflow/api_connexion/endpoints/task_instance_endpoint.py (hunk 4 above)
Comment (gyscos) [reply airflow/api_connexion/endpoints/task_instance_endpoint.py]: Correct, it shouldn't change anything (apart from making some static linters feel better)

File: airflow/api_connexion/parameters.py (hunk 3 above)
Comment (gyscos) [reply airflow/api_connexion/parameters.py]: Added

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 6]
```diff
@@ -85,17 +94,18 @@ def get_dag_run(dag_id, dag_run_id, session):
...
-    order_by='id',
+    session: Session,
+    dag_id: str,
+    start_date_gte: Optional[str] = None,
+    start_date_lte: Optional[str] = None,
+    execution_date_gte: Optional[str] = None,
+    execution_date_lte: Optional[str] = None,
+    end_date_gte: Optional[str] = None,
+    end_date_lte: Optional[str] = None,
+    state: Optional[List[str]] = None,
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
    state: Optional[List[DagRunState]] = None,
```

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 7]
```diff
@@ -124,16 +135,17 @@ def get_dag_runs(
...
-    offset,
-    order_by,
+    query: Query,
+    end_date_gte: Optional[str],
+    end_date_lte: Optional[str],
+    execution_date_gte: Optional[str],
+    execution_date_lte: Optional[str],
+    start_date_gte: Optional[str],
+    start_date_lte: Optional[str],
+    state: Optional[List[str]],
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
    state: Optional[List[DagRunState]],
```

File: airflow/api_connexion/endpoints/task_instance_endpoint.py (hunk 4 above)
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/task_instance_endpoint.py]:
```suggestion
    state: Optional[List[TaskInstanceState]] = None,
```

**Review by ephraimbuddy — COMMENTED** [review]
Looks good to me. Should we add a simple test too?

**Comment by uranusjr** [comment]
Yeah, tests are needed. Put them in `tests/api_connexion/endpoints/test_dag_run_endpoint.py`.

**Comment by ephraimbuddy** [comment]
Hi @gyscos , can you fix the conflicts and also add tests? Thanks

**Pull request closed without merging**